File: leetgo/go/LCR 123.cong-wei-dao-tou-da-yin-lian-biao-lcof/solution.go
// Created by k at 2024/05/16 15:45
// leetgo: dev
// https://leetcode.cn/problems/cong-wei-dao-tou-da-yin-lian-biao-lcof/

package main

import (
	"bufio"
	"fmt"
	"os"

	. "github.com/j178/leetgo/testutils/go"
)

// @lc code=begin
var ans []int

func reverseBookList(head *ListNode) []int {
	/* 1 */
	// for ; head != nil; head = head.Next {
	// 	ans = append([]int{head.Val}, ans...)
	// }

	/* 2 */
	// var pre *ListNode
	// for head != nil {
	// 	next := head.Next
	// 	head.Next = pre
	// 	pre = head
	// 	head = next
	// }
	// for ; pre != nil; pre = pre.Next {
	// 	ans = append(ans, pre.Val)
	// }
	ans = nil
	recur(head)
	return ans
}

func recur(head *ListNode) {
	if head == nil {
		return
	}
	recur(head.Next)
	ans = append(ans, head.Val)
}

// @lc code=end

func main() {
	stdin := bufio.NewReader(os.Stdin)
	head := Deserialize[*ListNode](ReadLine(stdin))
	ans := reverseBookList(head)

	fmt.Println("\noutput:", Serialize(ans))
}

File: leetgo/cpp/LCR 123.cong-wei-dao-tou-da-yin-lian-biao-lcof/solution.cpp
// Created by k at 2024/05/16 16:29
// leetgo: dev
// https://leetcode.cn/problems/cong-wei-dao-tou-da-yin-lian-biao-lcof/

#include <bits/stdc++.h>
#include "LC_IO.h"
using namespace std;

// @lc code=begin

class Solution {
  public:
    vector<int> reverseBookList(ListNode *head)
    {
        recur(head);
        return ans;
    }
    void recur(ListNode *head)
    {
        if (head == nullptr) {
            return;
        }
        recur(head->next);
        ans.push_back(head->val);
    }
    vector<int> ans;
};

// @lc code=end

int main()
{
    ios_base::sync_with_stdio(false);
    stringstream out_stream;

    ListNode *head;
    LeetCodeIO::scan(cin, head);

    Solution *obj = new Solution();
    auto res      = obj->reverseBookList(head);
    LeetCodeIO::print(out_stream, res);
    cout << "\noutput: " << out_stream.rdbuf() << endl;

    delete obj;
    return 0;
}
